src/app/digital/public/api/impl/DigitalComponent.ts

```diff
@@ -16,4 +18,34 @@ export class DigitalComponentImpl extends ComponentImpl<
     public override get info(): DigitalComponentInfo {
         return new DigitalComponentInfoImpl(this.circuit, this.kind);
     }
+
+    public override firstAvailable(portGroup: Port["group"]): DigitalPort | undefined {
+        const ports = this.internal.getPortsForComponent(this.id);
+
+        const inputTypes = this.info.inputPortGroups;
+        const outputTypes = this.info.outputPortGroups;
+
+        const isInputType = inputTypes.includes(portGroup);
+        const isOutputType = outputTypes.includes(portGroup)
+
+        if (!isInputType && !isOutputType) 
+            return undefined;
+
+        function firstAvailableHelper(comp: DigitalComponentImpl, id: string): Boolean {
+            let portObject = comp.internal.getPortByID(id);
+            let portWire = comp.internal.getWiresForPort(id); // no wires = available
+    
+            if (comp.info.inputPortGroups && portObject !== undefined && portObject.group === portGroup)
+                return true;  
+            if (comp.info.outputPortGroups && portObject !== undefined && portObject.group === portGroup && portWire.size === 0) {
+                return true;  
+            }
+            return false;
+        }
+
+        const match = Array.from(ports.values()).find(p => firstAvailableHelper(this, p));
```

Comment:
make changes to `firstAvailableHelper` then you can just do
`[...ports].find(firstAvailableHelper);`